File: lib/src/androidTest/java/com/pigdogbay/lib/utils/ObservablePropertyTest.java
package com.pigdogbay.lib.utils;


import android.support.test.filters.SmallTest;
import android.support.test.runner.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class ObservablePropertyTest {
    public enum Veg
    {
        carrots, peppers, onions, chillis, mushrooms
    }

    public class Listener implements ObservableProperty.PropertyChangedObserver<Veg>
    {
        public Veg newValue;

        @Override
        public void update(ObservableProperty<Veg> sender, Veg update) {
            newValue = update;
        }

    }
    @Test
    public void enum1()
    {
        ObservableProperty<Veg> target = new ObservableProperty<>(Veg.onions);
        Listener listener = new Listener();
        target.addObserver(listener);
        target.setValue(Veg.chillis);
        assertThat(listener.newValue, is(Veg.chillis));
    }


}

File: lib/src/test/java/com/pigdogbay/lib/patterns/ObservablePropertyTest.kt
package com.pigdogbay.lib.patterns

import org.junit.Test

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.nullValue
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThat


class ObservablePropertyTest : PropertyChangedObserver<ObservablePropertyTest.Veg> {

    var update = Veg.carrots
    var sender : Any? = null

    override fun update(sender: Any, update: Veg) {
        this.update = update
        this.sender = sender
    }

    enum class Veg {
        carrots, peppers, onions, chillis, mushrooms
    }

    @Test
    fun test1(){
        val obs = ObservableProperty(this, Veg.carrots)
        obs.addObserver(this)
        obs.value = Veg.chillis
        obs.removeObserver(this)
        assertThat(update, `is`(Veg.chillis))
        assertEquals(this,sender)
    }

    /**
     * No change, expect no update
     */
    @Test
    fun test2(){
        val obs = ObservableProperty(this, Veg.carrots)
        obs.addObserver(this)
        obs.value = Veg.carrots
        obs.removeObserver(this)
        assertThat(update, `is`(Veg.carrots))
        assertThat(sender, nullValue())
    }
}